File: framework/languages/java/e2e-kotlin/RegistrationCodec/Server/JsonOnlyPeer/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/jsononlypeer/Handlers/CodecHandlers.kt
package systems.zlink.e2e.kotlin.registrationcodec.jsononlypeer.handlers

import systems.zlink.e2e.kotlin.registrationcodec.protobuf.ProtobufEchoMsg
import systems.zlink.e2e.kotlin.registrationcodec.protobuf.ProtobufEchoReq
import systems.zlink.e2e.kotlin.registrationcodec.protobuf.ProtobufEchoRes
import systems.zlink.e2e.kotlin.registrationcodec.Contracts
import systems.zlink.e2e.kotlin.registrationcodec.JsonEchoMsg
import systems.zlink.e2e.kotlin.registrationcodec.JsonEchoReq
import systems.zlink.e2e.kotlin.registrationcodec.JsonEchoRes
import systems.zlink.e2e.kotlin.registrationcodec.PackedEchoMsg
import systems.zlink.e2e.kotlin.registrationcodec.PackedEchoRes
import systems.zlink.e2e.kotlin.registrationcodec.PackedEchoReq
import systems.zlink.e2e.kotlin.registrationcodec.jsononlypeer.infrastructure.ScenarioState
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.kotlin.ZLinkSuspendingSendHandler

@ZLinkHandlerGroup(Contracts.MANUAL_GROUP)
class JsonRequestHandler(
    private val state: ScenarioState,
) : ZLinkSuspendingRequestHandler<JsonEchoReq, JsonEchoRes> {
    override suspend fun handle(request: JsonEchoReq, context: ZLinkMessageContext): JsonEchoRes {
        state.record("Request", "JsonEchoReq", request.value)
        return JsonEchoRes("echo:${request.value}", "json")
    }
}

@ZLinkHandlerGroup(Contracts.MANUAL_GROUP)
class JsonSendHandler(
    private val state: ScenarioState,
) : ZLinkSuspendingSendHandler<JsonEchoMsg> {
    override suspend fun handle(message: JsonEchoMsg, context: ZLinkMessageContext) {
        state.record("Send", "JsonEchoMsg", message.value)
    }
}

@ZLinkHandlerGroup(Contracts.MANUAL_GROUP)
class ProtobufRequestHandler(
    private val state: ScenarioState,
) : ZLinkSuspendingRequestHandler<ProtobufEchoReq, ProtobufEchoRes> {
    override suspend fun handle(
        request: ProtobufEchoReq,
        context: ZLinkMessageContext,
    ): ProtobufEchoRes {
        state.record("Request", "ProtobufEchoReq", request.value)
        return ProtobufEchoRes.newBuilder().setValue("echo:${request.value}").build()
    }
}

@ZLinkHandlerGroup(Contracts.MANUAL_GROUP)
class ProtobufSendHandler(
    private val state: ScenarioState,
) : ZLinkSuspendingSendHandler<ProtobufEchoMsg> {
    override suspend fun handle(message: ProtobufEchoMsg, context: ZLinkMessageContext) {
        state.record("Send", "ProtobufEchoMsg", message.value)
    }
}

@ZLinkHandlerGroup(Contracts.MANUAL_GROUP)
class MsgpackRequestHandler(
    private val state: ScenarioState,
) : ZLinkSuspendingRequestHandler<PackedEchoReq, PackedEchoRes> {
    override suspend fun handle(request: PackedEchoReq, context: ZLinkMessageContext): PackedEchoRes {
        state.record("Request", "PackedEchoReq", request.value)
        return PackedEchoRes("echo:${request.value}")
    }
}

@ZLinkHandlerGroup(Contracts.MANUAL_GROUP)
class MsgpackSendHandler(
    private val state: ScenarioState,
) : ZLinkSuspendingSendHandler<PackedEchoMsg> {
    override suspend fun handle(message: PackedEchoMsg, context: ZLinkMessageContext) {
        state.record("Send", "PackedEchoMsg", message.value)
    }
}

File: framework/languages/dotnet/e2e/RegistrationCodec/Server/JsonOnlyPeer/Handlers/CodecHandlers.cs
using RegistrationCodec.Server.JsonOnlyPeer.Infrastructure;
using RegistrationCodec.Shared;
using Zlink.Framework.Contracts.Handlers;

namespace RegistrationCodec.Server.JsonOnlyPeer.Handlers;

internal sealed class JsonEchoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<JsonEchoReq, EchoRes>
{
    public ValueTask<EchoRes> HandleAsync(JsonEchoReq request, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-request|codec=json|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new EchoRes($"echo:{request.Value}", context.ContentType ?? "<null>"));
    }
}

internal sealed class JsonEchoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<JsonEchoMsg>
{
    public ValueTask HandleAsync(JsonEchoMsg message, IZLinkMessageContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add(
            $"codec-command|codec=json|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class ProtobufEchoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<ProtobufEchoReq, ProtobufEchoRes>
{
    public ValueTask<ProtobufEchoRes> HandleAsync(ProtobufEchoReq request, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-request|codec=protobuf|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(
            new ProtobufEchoRes { Value = $"echo:{request.Value}|content:{context.ContentType}" });
    }
}

internal sealed class ProtobufEchoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<ProtobufEchoMsg>
{
    public ValueTask HandleAsync(
        ProtobufEchoMsg message,
        IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-command|codec=protobuf|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class MessagePackEchoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<PackedEchoReq, PackedEchoRes>
{
    public ValueTask<PackedEchoRes> HandleAsync(PackedEchoReq request, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-request|codec=msgpack|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new PackedEchoRes
            { Value = $"echo:{request.Value}|content:{context.ContentType}" });
    }
}

internal sealed class MessagePackEchoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<PackedEchoMsg>
{
    public ValueTask HandleAsync(PackedEchoMsg message, IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add(
            $"codec-command|codec=msgpack|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}
